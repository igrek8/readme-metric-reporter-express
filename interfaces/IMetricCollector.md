[readme-metric-reporter-express](../README.md) / [Exports](../modules.md) / IMetricCollector

# Interface: IMetricCollector

## Table of contents

### Methods

- [collect](IMetricCollector.md#collect)

## Methods

### collect

â–¸ **collect**(`req`, `res`, `metric`): `Metric`

#### Parameters

| Name | Type |
| :------ | :------ |
| `req` | `Request`<`ParamsDictionary`, `any`, `any`, `ParsedQs`, `Record`<`string`, `any`\>\> |
| `res` | `Response`<`any`, `Record`<`string`, `any`\>\> |
| `metric` | `Metric` |

#### Returns

`Metric`

#### Defined in

[IMetricCollector.ts:5](https://github.com/igrek8/readme-metric-reporter-express/blob/b3d4e3b/src/IMetricCollector.ts#L5)
