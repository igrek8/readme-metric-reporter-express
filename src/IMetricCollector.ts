import { Request, Response } from 'express';
import { Metric } from 'readme-metric-reporter';

export interface IMetricCollector {
  collect(req: Request, res: Response, metric: Metric): Metric;
}
