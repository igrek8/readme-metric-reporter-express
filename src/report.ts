import { randomUUID } from 'crypto';
import type { NextFunction, Request, Response } from 'express';
import { release, type } from 'os';
import {
  IMetricReporter,
  MetricRequestHeader,
  MetricRequestQueryString,
  MetricResponseHeader,
} from 'readme-metric-reporter';
import { URL } from 'url';

import { IMetricCollector } from './IMetricCollector';
import { serialize } from './utils/serialize';

// eslint-disable-next-line @typescript-eslint/no-var-requires
const { name, version } = require('../package.json');

export interface ReportOptions {
  /**
   * Base URL to your readme.io project
   */
  baseUrl?: string;
  reporter: IMetricReporter;
  collector: IMetricCollector;
  /**
   * Create a metric id for the request
   */
  createUUID?(req: Request, res: Response): string;
}

const comment = `${type()}/${release()}`.toLowerCase();

const EXPLORER_HEADER = 'x-readme-api-explorer';

const LOG_URL_HEADER = 'x-documentation-url';

export function report({ collector, reporter, createUUID = () => randomUUID(), baseUrl }: ReportOptions) {
  return (req: Request, res: Response, next: NextFunction) => {
    const start = new Date();
    const _id = createUUID(req, res);
    const isExplorer = typeof req.header(EXPLORER_HEADER) === 'string';

    /**
     * Capture outgoing response body
     */
    let _body: unknown;
    const _send = res.send;
    res.send = function send(body?: unknown) {
      _body = body;
      res.send = _send;
      return res.send(body);
    };

    if (isExplorer && typeof baseUrl === 'string') {
      const url = new URL(baseUrl);
      url.pathname = `/logs/${_id}`;
      res.setHeader(LOG_URL_HEADER, url.toString());
    }

    res.once('finish', async () => {
      try {
        const end = new Date();
        const elapsed = end.getTime() - start.getTime();
        const url = req.protocol + '://' + req.get('host') + req.originalUrl;
        const httpVersion = `${req.protocol.toUpperCase()}/${req.httpVersion}`;
        const queryString = Object.entries(req.query).reduce<MetricRequestQueryString[]>((arr, [name, value]) => {
          if (value) {
            arr.push({
              name,
              value: value.toString(),
            });
          }
          return arr;
        }, []);
        const requestHeaders = Object.entries(req.headers).reduce<MetricRequestHeader[]>((arr, [name, value]) => {
          if (value) {
            arr.push({
              name,
              value: value.toString(),
            });
          }
          return arr;
        }, []);
        const responseHeaders = Object.entries(res.getHeaders()).reduce<MetricResponseHeader[]>(
          (arr, [name, value]) => {
            if (value) {
              arr.push({
                name,
                value: value.toString(),
              });
            }
            return arr;
          },
          []
        );
        await reporter.report(
          collector.collect(req, res, {
            _id,
            group: {
              id: req.ip,
            },
            clientIPAddress: req.ip,
            request: {
              log: {
                creator: {
                  name,
                  version,
                  comment,
                },
                entries: [
                  {
                    startedDateTime: end.toISOString(),
                    time: elapsed,
                    pageref: url,
                    request: {
                      url,
                      method: req.method,
                      headers: requestHeaders,
                      httpVersion: httpVersion,
                      queryString: queryString,
                      postData: {
                        mimeType: req.header('content-type')?.toString(),
                        text: serialize(req.body),
                      },
                    },
                    response: {
                      status: res.statusCode,
                      statusText: res.statusMessage,
                      headers: responseHeaders,
                      content: {
                        mimeType: req.header('content-type')?.toString(),
                        size: req.socket.bytesWritten,
                        text: serialize(_body),
                      },
                    },
                  },
                ],
              },
            },
          })
        );
        /* c8 ignore next */
      } catch {} // eslint-disable-line no-empty
    });
    next();
  };
}
