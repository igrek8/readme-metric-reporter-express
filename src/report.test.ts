import { Request, Response } from 'express';
import { Server } from 'http';
import { Metric, MetricReporter } from 'readme-metric-reporter';
import * as request from 'supertest';
import { v4, validate } from 'uuid';
import express = require('express');

import { IMetricCollector } from './IMetricCollector';
import { report } from './report';

jest.useFakeTimers();
jest.setSystemTime(new Date('2022-01-01'));

jest.mock('os', () => ({
  ...jest.requireActual('os'),
  type: jest.fn().mockReturnValue('darwin'),
  release: jest.fn().mockReturnValue('10.0.0'),
}));

jest.mock('uuid');

const reporter = new MetricReporter('apiKey');

const collector: IMetricCollector = {
  collect(_req: Request, _res: Response, metric: Metric) {
    return metric;
  },
};

describe('createMiddleware', () => {
  let server: Server;

  beforeAll((done) => {
    server = express()
      .use(report({ reporter, collector, baseUrl: 'http://example.com' }))
      .use(express.json())
      .use((req, res) => res.status(200).json({ echo: req.body }))
      .listen(50000, done);
  });

  afterAll((done) => {
    server.close(done);
  });

  it('reports metric', async () => {
    (v4 as jest.MockedFunction<typeof v4>).mockReturnValueOnce('00000000-0000-0000-0000-000000000000');
    (validate as jest.MockedFunction<typeof validate>).mockReturnValueOnce(true);
    const report = jest.spyOn(reporter, 'report').mockImplementation(async () => {});
    await request(server).post('/echo?search=criteria').send({ message: 'test' }).set('x-readme-api-explorer', '0.0.0');
    expect(report.mock.calls[0]?.[0]).toMatchSnapshot();
  });
});
