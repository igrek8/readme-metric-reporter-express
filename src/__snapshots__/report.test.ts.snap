// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createMiddleware reports metric 1`] = `
{
  "_id": "00000000-0000-0000-0000-000000000000",
  "clientIPAddress": "::ffff:127.0.0.1",
  "group": {
    "id": "::ffff:127.0.0.1",
  },
  "request": {
    "log": {
      "creator": {
        "comment": "darwin/10.0.0",
        "name": "readme-metric-reporter-express",
        "version": "0.0.0",
      },
      "entries": [
        {
          "pageref": "http://127.0.0.1:50000/echo?search=criteria",
          "request": {
            "headers": [
              {
                "name": "host",
                "value": "127.0.0.1:50000",
              },
              {
                "name": "accept-encoding",
                "value": "gzip, deflate",
              },
              {
                "name": "content-type",
                "value": "application/json",
              },
              {
                "name": "x-readme-api-explorer",
                "value": "0.0.0",
              },
              {
                "name": "content-length",
                "value": "18",
              },
              {
                "name": "connection",
                "value": "close",
              },
            ],
            "httpVersion": "HTTP/1.1",
            "method": "POST",
            "postData": {
              "mimeType": "application/json",
              "text": "{"message":"test"}",
            },
            "queryString": [
              {
                "name": "search",
                "value": "criteria",
              },
            ],
            "url": "http://127.0.0.1:50000/echo?search=criteria",
          },
          "response": {
            "content": {
              "mimeType": "application/json",
              "size": 317,
              "text": "{"echo":{"message":"test"}}",
            },
            "headers": [
              {
                "name": "x-powered-by",
                "value": "Express",
              },
              {
                "name": "x-documentation-url",
                "value": "http://example.com/logs/00000000-0000-0000-0000-000000000000",
              },
              {
                "name": "content-type",
                "value": "application/json; charset=utf-8",
              },
              {
                "name": "content-length",
                "value": "27",
              },
              {
                "name": "etag",
                "value": "W/"1b-VUQuMVMiOaIDRK80a/GKMM7BGtI"",
              },
            ],
            "status": 200,
            "statusText": "OK",
          },
          "startedDateTime": "2022-01-01T00:00:00.000Z",
          "time": 0,
        },
      ],
    },
  },
}
`;
